Add ml-commons passthrough post process function

q-andy opened the pull request:
### Description
Adds a predefined post-process function to unwrap model output when making a remote call to second ml-commons cluster. When calling a remote second ml-commons predict API with a remote connector, the remote connector output will "double wrap" the second ml-commons. Tested compatibility with neural-search `SparseEncodingProcessor` and `TextEmbeddingProcessor`.


Predict call without post process function (double wrapped):
```
{
	"inference_results": [
		{
			"output": [
				{
					"name": "output",
					"dataAsMap": {
						"inference_results": [
							{
								"output": [
									{
										"name": "output",
										"dataAsMap": {
											"response": [
												{
													"this": 0.6228184700012207,
													"harrison": 0.5467907786369324,
													...
												},
												{
													"?": 0.1258760392665863,
													"day": 0.46702781319618225,
													...
												}
											]
										}
									}
								],
								"status_code": 200.0
							}
						]
					}
				}
			],
			"status_code": 200
		}
	]
}
```

After using post process function:
```
{
	"inference_results": [
		{
			"output": [
				{
					"name": "output",
					"dataAsMap": {
						"response": [
							{
								"increasingly": 0.028670792,
								"achievements": 0.4906937,
								...
							}
						]
					}
				},
				{
					"name": "output",
					"dataAsMap": {
						"response": [
							{
								"hesse": 0.07568395,
								"greeting": 0.296827,
								...
							}
						]
					}
				}
			],
			"status_code": 200
		}
	]
}
```

### Related Issues
Resolves #[Issue number to be closed when this PR is merged]

### Check List
- [x] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md).
- [x] Commits are signed per the DCO using `--signoff`.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/ml-commons/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

Zhangxunmt commented on common/src/main/java/org/opensearch/ml/common/connector/functions/postprocess/RemoteMlCommonsPassthroughPostProcessFunction.java:
```diff
@@ -0,0 +1,117 @@
+     * of the /_predict API
+     */
+    @SuppressWarnings("unchecked")
+    private ModelTensor createModelTensorFromMap(Map<String, Object> map) {
```
should add validation for the map to be not null and some fields exist.

Zhangxunmt commented on common/src/main/java/org/opensearch/ml/common/connector/functions/postprocess/RemoteMlCommonsPassthroughPostProcessFunction.java:
```diff
@@ -0,0 +1,117 @@
+        MLResultDataType dataType = null;
+        if (map.containsKey(ModelTensor.DATA_TYPE_FIELD)) {
+            Object dataTypeObj = map.get(ModelTensor.DATA_TYPE_FIELD);
+            if (dataTypeObj instanceof String) {
```
nitpick: There are multiple unchecked type casting operations in this method. I think it worth adding type safety to avoid too much "unchecked" warnings.

dhrubo-os commented:
@q-andy fix spotless :)

q-andy commented:
Most CI passed, just 2 flaky integ test failures are unrelated to this change. Could you take another look @dhrubo-os @Zhangxunmt

q-andy commented on common/src/main/java/org/opensearch/ml/common/connector/functions/postprocess/RemoteMlCommonsPassthroughPostProcessFunction.java:
```diff
@@ -0,0 +1,121 @@
+            if (dataTypeObj instanceof String) {
+                try {
+                    dataType = MLResultDataType.valueOf((String) dataTypeObj);
+                } catch (IllegalArgumentException e) {
```
I left as null instead of treating like 5xx is because even if the model data type is invalid, there may be usecases where we can still parse and use the model data or dataAsMap. E.g. some model types like neural sparse or NER don't include data type as part of the model response, so null is still valid.

Right now since we're focussed on neural sparse and dense, my thought process is its better to leave it flexible to be able to possible handle different model response formats. Updated the comment to explain this.

q-andy commented on common/src/main/java/org/opensearch/ml/common/connector/functions/postprocess/RemoteMlCommonsPassthroughPostProcessFunction.java:
```diff
@@ -0,0 +1,121 @@
+        Number[] data = null;
+        if (map.containsKey(ModelTensor.DATA_FIELD)) {
+            Object dataObj = map.get(ModelTensor.DATA_FIELD);
+            if (dataObj instanceof List<?> dataList) {
```
See above comment, this is valid for neural sparse and other remote model format. Looks like data field is primarily used for vector/numerical info, and if a model output doesn't include that, then it uses dataAsMap instead and data field being null is valid.

q-andy commented:
>the name seems changed, the name used to be "response" but after the post processing become "output". Is this change intended? I am sure if the name is used somewhere but better not change the name if not intentionally.

Fixed, the name is relevant for different model types. I changed it so the name will be passthroughed as well, this is just a typo in the PR description.

q-andy commented:
Could we backport to 2.19 and 3.1? For 2.19 back port there will be some conflicts, I will resolve

cc: @dhrubo-os @Zhangxunmt
